src/video.c

```diff
@@ -8,44 +10,203 @@ _XL_OBJECTTYPE gClassVideo = {
-    &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008EC98, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008ECB4,
-    &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008ECC0, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008EDE8,
-    &lbl_8008ECD0, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008ECE0, &lbl_8008EDE8, &lbl_8008EDE8,
-    &lbl_8008EDE8, &lbl_8008ECF0, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008ED38, &lbl_8008EDE8,
-    &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008ED50, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008ED68,
-    &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008EDE8, &lbl_8008EDA8,
-};
+s32 videoPut16(Video* pVideo, u32 nAddress, s16* pData) { return 0; }
+
+s32 videoPut32(Video* pVideo, u32 nAddress, s32* pData) {
+    void* pRAM;
+    Frame* pFrame;
+    FrameBuffer* pBuffer;
+
+    switch (nAddress & 0x3F) {
+        case 0x0:
+            pVideo->nStatus = *pData & 0xFFFF;
+            break;
+        case 0x4:
+            pVideo->nAddress = *pData & 0xFFFFFF;
+            pFrame = SYSTEM_FRAME(pVideo->pHost);
+            pBuffer = &pFrame->aBuffer[2];
+
+            if (!ramGetBuffer(SYSTEM_RAM(pVideo->pHost), &pRAM, pVideo->nAddress, NULL)) {
+                return 0;
+            }
+
+            if (pBuffer->pData != pRAM) {
+                pBuffer->nFormat = 0;
+                pBuffer->nSize = 2;
+                pBuffer->nWidth = pVideo->nSizeX;
+                pBuffer->pData = pRAM;
+
+                if (!frameSetBuffer(pFrame, 2)) {
+                    return 0;
+                }
+            }
+            break;
+        case 0x8:
+            pVideo->nSizeX = *pData & 0xFFF;
+            break;
+        case 0xC:
+            pVideo->nScanInterrupt = *pData & 0x3FF;
+            break;
+        case 0x10:
+            xlObjectEvent(pVideo->pHost, 0x1001, (void*)8);
+            pVideo->nScanInterrupt = 0x10000;
+            break;
+        case 0x14:
+            pVideo->nTiming = *pData;
+            break;
+        case 0x18:
+            pVideo->nSyncV = *pData & 0x3FF;
+            break;
+        case 0x1C:
+            pVideo->nSyncH = *pData & 0x1FFFFF;
+            break;
+        case 0x20:
+            pVideo->nSyncLeap = *pData & 0x0FFFFFFF;
+            break;
+        case 0x24:
+            if ((pVideo->nStartH = *pData & 0x03FF03FF) == 0) {
+                if (pVideo->bBlack != 1) {
+                    pVideo->bBlack = 1;
+                }
+            } else if (pVideo->bBlack) {
+                pVideo->bBlack = 0;
+            }
+            break;
+        case 0x28:
+            pVideo->nStartV = *pData & 0x03FF03FF;
+            break;
+        case 0x2C:
+            pVideo->nBurst = *pData & 0x03FF03FF;
+            break;
+        case 0x30:
+            pVideo->nScaleX = *pData & 0xFFF;
+            if (!frameSetSize(SYSTEM_FRAME(pVideo->pHost), 0, pVideo->nSizeX, (s32)(pVideo->nScaleY * 0xF0) / 1024)) {
```

Comment:
```suggestion
            if (!frameSetSize(SYSTEM_FRAME(pVideo->pHost), 0, pVideo->nSizeX, (s32)(pVideo->nScaleY * 240) / 1024)) {
```
ditto below (recognizable as 240p N64 screen size)